### PR TITLE
feat(payout): add dynamic fields for payout links

### PR DESCRIPTION
**Pull request by kashif-m**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
This PR includes below changes
- add dynamic fields support for payout widget (#5765)
- expose `formLayout` option in payout link config - journey or tabs

### Additional Changes

- [x] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [x] This PR modifies application configuration/environment variables

## Motivation and Context
Allows for collecting data dynamically based on the selected payout method type.

## How did you test it?
Tested locally

<details>
    <summary>1. Create a payout widget which uses journey layout</summary>

    curl --location 'http://localhost:8080/payouts/create' \
    --header 'Content-Type: application/json' \
    --header 'api-key: dev_6lYcIjzguBUudkK9bpoM3LZDod7RdE1wDJWzxwZhGLWWL6mlQIcfbUkXQgTAgj7e' \
    --data '{
        "amount": 1,
        "currency": "EUR",
        "customer_id": "cus_s5yZDiR680XOcGjcZrQ4",
        "description": "Its my first payout request",
        "entity_type": "Individual",
        "confirm": false,
        "auto_fulfill": true,
        "session_expiry": 1000000,
        "priority": "instant",
        "profile_id": "pro_zFe41ju6ZVZMtuuUHPji",
        "payout_link": true,
        "return_url": "https://www.google.com",
        "payout_link_config": {
            "formLayout": "journey",
            "theme": "#0066ff"
        }
    }'

image

</details>

<details>
    <summary>2. Create a payout widget which uses tabs layout</summary>

    curl --location 'http://localhost:8080/payouts/create' \
    --header 'Content-Type: application/json' \
    --header 'api-key: dev_6lYcIjzguBUudkK9bpoM3LZDod7RdE1wDJWzxwZhGLWWL6mlQIcfbUkXQgTAgj7e' \
    --data '{
        "amount": 1,
        "currency": "EUR",
        "customer_id": "cus_s5yZDiR680XOcGjcZrQ4",
        "description": "Its my first payout request",
        "entity_type": "Individual",
        "confirm": false,
        "auto_fulfill": true,
        "session_expiry": 1000000,
        "priority": "instant",
        "profile_id": "pro_zFe41ju6ZVZMtuuUHPji",
        "payout_link": true,
        "return_url": "https://www.google.com",
        "payout_link_config": {
            "formLayout": "tabs",
            "theme": "#0066ff"
        }
    }'

image

</details>

<details>
    <summary>3. Create a payout without billing address</summary>

    curl --location 'http://localhost:8080/payouts/create' \
    --header 'Content-Type: application/json' \
    --header 'api-key: dev_6lYcIjzguBUudkK9bpoM3LZDod7RdE1wDJWzxwZhGLWWL6mlQIcfbUkXQgTAgj7e' \
    --data '{
        "amount": 1,
        "currency": "EUR",
        "customer_id": "cus_s5yZDiR680XOcGjcZrQ4",
        "description": "Its my first payout request",
        "entity_type": "Individual",
        "confirm": false,
        "auto_fulfill": true,
        "session_expiry": 1000000,
        "priority": "instant",
        "profile_id": "pro_zFe41ju6ZVZMtuuUHPji",
        "payout_link": true,
        "return_url": "https://www.google.com",
        "payout_link_config": {
            "formLayout": "tabs",
            "theme": "#0066ff"
        }
    }'

image

</details>

<details>
    <summary>4. Create a payout with partial billing address</summary>

    curl --location 'http://localhost:8080/payouts/create' \
    --header 'Content-Type: application/json' \
    --header 'api-key: dev_6lYcIjzguBUudkK9bpoM3LZDod7RdE1wDJWzxwZhGLWWL6mlQIcfbUkXQgTAgj7e' \
    --data '{
        "amount": 1,
        "currency": "EUR",
        "customer_id": "cus_s5yZDiR680XOcGjcZrQ4",
        "description": "Its my first payout request",
        "entity_type": "Individual",
        "confirm": false,
        "auto_fulfill": true,
        "session_expiry": 1000000,
        "priority": "instant",
        "profile_id": "pro_zFe41ju6ZVZMtuuUHPji",
        "payout_link": true,
        "return_url": "https://www.google.com",
        "payout_link_config": {
            "formLayout": "tabs",
            "theme": "#0066ff"
        }
    }'

image

</details>

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/router/src/core/payouts/transformers.rs [hunk 1]
```diff
@@ -103,3 +110,78 @@ impl
...
+                            .and_then(|pmt_info| {
+                                pmt_info.0.get(&pmt).map(|connector_fields| {
+                                    let mut required_fields = IndexMap::new();
+
+                                    for required_field_final in connector_fields.fields.values() {
+                                        required_fields.extend(required_field_final.common.clone());
+                                        required_fields
+                                            .extend(required_field_final.mandate.clone());
+                                        required_fields
+                                            .extend(required_field_final.non_mandate.clone());
```
Comment (Chethan-rao) [inline crates/router/src/core/payouts/transformers.rs:150]: is this expected to extend required_fields with both mandate and non_mandate fields without any validation?

File: crates/router/src/core/payouts/transformers.rs [hunk 2]
```diff
@@ -103,3 +110,78 @@ impl
...
+                let payment_method_types_info = enabled_payout_method
+                    .payment_method_types
+                    .into_iter()
+                    .filter_map(|pmt| {
+                        payout_required_fields
+                            .0
+                            .get(&payment_method)
+                            .and_then(|pmt_info| {
+                                pmt_info.0.get(&pmt).map(|connector_fields| {
+                                    let mut required_fields = IndexMap::new();
```
Comment (Chethan-rao) [inline crates/router/src/core/payouts/transformers.rs:143]: any specific reason for using IndexMap?

**Review by Chethan-rao — CHANGES_REQUESTED** [review]

File: crates/router/src/core/payouts/transformers.rs (hunk 1 above)
Comment (kashif-m) [reply crates/router/src/core/payouts/transformers.rs:150]: There is no mandate concept in payouts, so we can completely remove the extension to `mandate` and `non_mandate` fields

File: crates/router/src/core/payouts/transformers.rs (hunk 2 above)
Comment (kashif-m) [reply crates/router/src/core/payouts/transformers.rs:143]: Wanted to keep the elements in order while serialisation, this didn't work anyway. Will revert it back to HashMap 👍

File: crates/router/Cargo.toml [hunk 3]
```diff
@@ -70,6 +70,7 @@ hex = "0.4.3"
 http = "0.2.12"
 hyper = "0.14.28"
 image = { version = "0.25.1", default-features = false, features = ["png"] }
+indexmap = "2.2.6"
```
Comment (Chethan-rao) [inline crates/router/Cargo.toml:73]: You can remove this dependency addition too?

**Review by Chethan-rao — APPROVED** [review]

File: crates/router_derive/src/lib.rs [hunk 4]
```diff
@@ -615,3 +613,85 @@ pub fn try_get_enum_variant(input: proc_macro::TokenStream) -> proc_macro::Token
...
+///
+/// [
+///     ("name", "Test"),
+///     ("address.line1", "1397"),
+///     ("address.line2", "Some street"),
+///     ("address.zip", "941222"),
+///     ("email", "test@example.com"),
+/// ]
+#[proc_macro_derive(FlattenStructKeys)]
+pub fn flatten_struct_keys_derive(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
```
Comment (jarnura) [inline crates/router_derive/src/lib.rs:649]: Add unit test cases for this function

**Review by jarnura — APPROVED** [review]

File: crates/router_derive/src/lib.rs (hunk 4 above)
Comment (kashif-m) [reply crates/router_derive/src/lib.rs]: Added an unit test case for this macro

**Review by Chethan-rao — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

**Pull request merged**